[IMP] html_editor, project: add link preview new feature

Jinjiu96 opened the pull request:
Description
-----------------
This commit introduces the new feature to preview link elements upon
clicking and assigns customized link titles for project.task.

Implementation
-----------------
External Links: The preview metadata for external links is retrieved
using an existing method from the mail module. If the external link is
pointing to an image, we display it in the preview popover and title it
with the link label.

Internal Links: To display customized titles for internal links, a
computed field named link_preview_name is introduced. This allows
displaying titles in a format specific to the model, such as
"TaskName | ProjectName" for tasks within projects. The
link_preview_name field is model-specific and is determined by the
preferred data representation for the preview. In cases where
link_preview_name is not defined, the system falls back to display_name,
and subsequently to name, or ultimately uses the webpage's title if none
of these fields are available.

When the displayed link label differs from the preview title, a banner
is displayed at the bottom of the preview indicating the change to the
preview title.

Error Handling
-----------------
The approach to error handling is designed to ensure that the preview
functionality does not interrupt the user flow.

External Links: If metadata retrieval fails for an external link, the
preview is removed, and only the favicon along with the URL is displayed

Internal Links: For internal links, specific errors such as value or
index errors indicate incorrect record links, triggering a warning
notification. Other types of errors are logged to the front-end console
without raising an exception in the controller, to prevent any
disruption to the user interface.


task-3678777

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/174505.png)](https://mergebot.odoo.com/odoo/odoo/pull/174505)

dmo-odoo commented:
@Jinjiu96 Is it https://www.odoo.com/odoo/project/1695/tasks/3678777?cids=1 or something else?

Jinjiu96 commented:
> @Jinjiu96 Is it https://www.odoo.com/odoo/project/1695/tasks/3678777?cids=1 or something else?

@dmo-odoo yes exactly! I was working on the new feature tasks of links :smile:  For the review, I'm not very sure how do we want to process it. I posted in the phoenix channel directly cause it's based on the phoenix but didn't think about the Indian test phase ^^'

Shall we put it into our normal reviewing pipe?

dmo-odoo commented:
Could you add a task id to the commit and PR description please? :)

fdardenne commented on addons/html_editor/controllers/main.py:
```diff
@@ -547,3 +562,47 @@ def bus_broadcast(self, model_name, field_name, res_id, bus_data):
         bus_data.update({'model_name': model_name, 'field_name': field_name, 'res_id': res_id})
         request.env['bus.bus']._sendone(channel, 'editor_collaboration', bus_data)
 
+    @http.route('/html_editor/link_preview_external', type="json", auth="public", methods=['POST'])
```
I'm surprised we need this route, I may be wrong but I think it's best for the client to fetch the metadata himself.

Jinjiu96 replied:
@fdardenne I tried to fetch the external links directly in the frontend at the beginning, but usually it receives an error because of the CORS policy. That's why I used the back-end for it. But maybe there's a way to do it at the front-end properly? :pray: 

@xmo-odoo If it's necessary and should be in `mail`, shall I create a new file in the mail controller and add it? Sorry for the naive question, just wanna be careful when touching another model :pray:

fdardenne replied:
My bad @Jinjiu96, you were right. I just tested how other known social network does and they do not let the browser fetch the preview. 
Security wise it seems logic after all, we don't want Bob to put his "malicous" website in the editor and get all the browser fingerprint informations about each reader previewing the link.

fdardenne commented on addons/html_editor/controllers/main.py:
```diff
@@ -547,3 +562,47 @@ def bus_broadcast(self, model_name, field_name, res_id, bus_data):
+    def link_preview_metadata(self, preview_url):
+        return link_preview.get_link_preview_from_url(preview_url)
+
+    @http.route('/html_editor/link_preview_internal', type="json", auth="public", methods=['POST'])
```
Do we need it to be public ?

mart-e replied:
especially if we fetch the action in sudo

Jinjiu96 replied:
Yes, I agree it'll be better to be `user`. I made it public cause I thought since the person already has the link, they should have the right to preview it even though not authenticated. But I didn't think about the sudo action fetch :pray:

fdardenne commented on addons/html_editor/static/src/main/link/link_popover.js:
```diff
@@ -183,34 +192,80 @@ export class LinkPopover extends Component {
+                .add(fetch(this.state.url))
                 .then((response) => response.text())
-                .then((content) => {
-                    const parser = new window.DOMParser();
-                    const doc = parser.parseFromString(content, "text/html");
+                .then(async (content) => {
+                    const html_parser = new window.DOMParser();
+                    const doc = html_parser.parseFromString(content, "text/html");
 
                     // Get
                     const favicon = doc.querySelector("link[rel~='icon']");
                     const ogTitle = doc.querySelector("[property='og:title']");
                     const title = doc.querySelector("title");
+                    // Get the metadata internally
+                    const internalUrlData = await rpc("/html_editor/link_preview_internal", {
+                        preview_url: url.pathname,
+                    });
```
Can't we fetch the metadata in 1 call without calling the backend here ?

fdardenne reviewed (commented):
I'm not very familiar with the backend side but here are my remarks. 
I see that there is a public route with a `safe_eval` in there, it would be prudent to let @odoo/rd-security assess this. And do we really need it to be public ?

Otherwise good job, it's a nice feature 👍

mart-e commented on addons/html_editor/controllers/main.py:
```diff
@@ -547,3 +562,47 @@ def bus_broadcast(self, model_name, field_name, res_id, bus_data):
+                return model.browse(int(record_id)).web_read(spec)[0]
+            return {}
+        except (IndexError, ValueError) as e:
+            return {'error_msg': _(str(e) + ", link preview is not available, please check your url is correct or if the model has the link preview field")}
```
No variables inside gettext calls
https://www.odoo.com/documentation/master/developer/howtos/translations.html#variables

Jinjiu96 commented:
Thanks a lot for the reviews guys! I wasn't very familiar with writing routes so there are many places to improve, I will adapt it based on the comments :pray:

xavierbol commented on addons/project/models/project_task.py:
```diff
@@ -779,6 +780,13 @@ def _inverse_display_name(self):
                         extract_data(task)
                 task.name = task.display_name.strip()
 
+    def _compute_link_preview_name(self):
+        for task in self:
+            if task.project_id.id:
+                task.link_preview_name = task.display_name + ' | ' + task.project_id.name
+            else:
+                task.link_preview_name = task.display_name
```
Seems equivalent to:
```suggestion
    def _compute_link_preview_name(self):
        for task in self:
            link_preview_name = task.display_name
            if task.project_id:
                link_preview_name += f' | {task.project_id.name}'
            task.link_preview_name = link_preview_name
```
Also, the user can have access to the task but no always to the project. So, we should add a sudo to get the project name.

```suggestion
    def _compute_link_preview_name(self):
        for task in self:
            link_preview_name = task.display_name
            if task.project_id:
                link_preview_name += f' | {task.project_id.sudo().name}'
            task.link_preview_name = link_preview_name
```

xavierbol commented on addons/project/models/project_task.py:
```diff
@@ -299,6 +299,7 @@ def _read_group_personal_stage_type_ids(self, stages, domain):
             ! Set the task a high priority\n
             Make sure to use the right format and order e.g. Improve the configuration screen #feature #v16 @Mitchell !""",
     )
+    link_preview_name = fields.Char(compute='_compute_link_preview_name', help="Name used in the link preview")
```
It is more a technical field, no? If yes, then we don't really need to add an helper message, IMO.

Also, we can also avoid translating the string for this techincal field by adding `export_string_translation=False` inside the field definition. 🤷🏼‍♂️
(except if we display the label of this field somewhere) 

```suggestion
    link_preview_name = fields.Char(compute='_compute_link_preview_name', export_string_translation=False)
```

xavierbol reviewed (commented):
Thanks for your work, I left 2 comments. 

(I just review the changes made in project module)

Jinjiu96 commented:
Hello guys, I updated the PR and added two more JS unit tests. @xmo-odoo I kept the routes in the html_editor cause the route is only used in html editor. I talked with someone in discuss about moving it to `mail/controllers/link_preview.py`, but they think it's better to keep it in html_editor where the routes are used.

Thanks again for reviewing and it really helped a lot to improve the code. Please feel free to let me know if there's something else to change :pray: If it looks good to you guys, I will squash the commits :smile:

xavierbol reviewed (approved):
LGTM for the services part, thanks for your work. 🙂

Jinjiu96 commented on addons/html_editor/controllers/main.py:
```diff
@@ -547,3 +549,45 @@ def bus_broadcast(self, model_name, field_name, res_id, bus_data):
+            action = Actions.sudo().search([('path', '=', action_name)])
+            if not action:
+                return {'error_msg': _("Action %s not found, link preview is not available, please check your url is correct", action_name)}
+            action_type = action.type
+            action = request.env[action_type].browse(action.id)
+            context.update(safe_eval(action.context, dict(
```
Thanks a bunch for re-checking it! I didn't think very deeply about the action type. When it's not a window action, the record reading will fall into the all exception catch. But I think you are right that we should only run the code when it's a window action. For the context updating, I didn't know it's only the lang matters for a computed field.

I will update the code and remove the context updating :pray:

Jinjiu96 commented on addons/html_editor/controllers/main.py:
```diff
@@ -547,3 +549,45 @@ def bus_broadcast(self, model_name, field_name, res_id, bus_data):
+            model = request.env[action.res_model].with_context(context)
+            record = model.browse(record_id)
+
+            result = {}
+            if hasattr(model, 'description'):
+                result['description'] = record.description
+            if hasattr(model, 'link_preview_name'):
+                result['link_preview_name'] = record.link_preview_name
+            elif hasattr(model, 'display_name'):
+                result['display_name'] = record.display_name
+            elif hasattr(model, 'name'):
+                result['name'] = record.name
+
+            return result
```
Thanks! I will adapt it :pray:

Jinjiu96 commented:
@xmo-odoo Thanks a bunch! I added it and quashed the commits :pray:

fdardenne reviewed (approved):
@robodoo r+

fdardenne commented:
@robodoo rebase-ff

robodoo commented:
@Jinjiu96 @fdardenne because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

robodoo commented:
Merge method set to rebase and fast-forward.
